local from scratch: not using llamaindex or langchain

# Librairies Installation

In [ ]:
import os

if "COLAB_GPU" in os.environ:
    print("[INFO] Running in Google Colab, installing requirements.")
    !pip install -U torch
    !pip install PyMuPDF
    !pip install tqdm
    !pip install sentence-transformers
    !pip install accelerate
    !pip install bitsandbytes
    !pip install flash-attn --no-build-isolation

[INFO] Running in Google Colab, installing requirements.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 42.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.9/15.9 MB 66.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 245.3/245.3 kB 15.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.5/137.5 MB 7.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 81.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for flash-attn: filename=flash_attn-2.6.3-cp310-cp310-linux_x86_64.whl size=187309225 sha256=237ef9c6157db394e1ddde4ba609a21ebb98382377a27041edc09318801a6f24
  Stored in directory: /root/.cache/pip/wheels/7e/e3/c3/89c7a2f3c4adc07cd1c675f8bb7b9ad4d18f64a72bccdfe826
Successfully built flash-attn


# PDF Data Processing

The process is as follows:
1. Break the doc into pages
2. Break the pages into sentences
3. Break the sentences into chunks (To prevent overloading the embedding models capacity for tokens)

Tokens can be thought of as pieces of words. Before the API processes the request, the input is broken down into tokens. These tokens are not cut up exactly where the words start or end - tokens can include trailing spaces and even sub-words. Here are some helpful rules of thumb for understanding tokens in terms of lengths:

1. **1 token ~= 4 chars in English**
2. 1 token ~= ¾ words
3. 100 tokens ~= 75 words

In [ ]:
doc_path='/content/nutrition.pdf'
import pymupdf
from tqdm import tqdm
def open_and_read_pdf(pdf_path):
  def format(text) :
    cleaned_text = text.replace("\n", " ").strip()
    cleaned_text = " ".join(cleaned_text.split()) #remove multiple spaces
    cleaned_text = cleaned_text.lower()
    return cleaned_text
  doc = pymupdf.open(doc_path)
  pages_and_texts = []
  for page_number, page in tqdm(enumerate(doc)):
        text = page.get_text()
        formatted_text = format(text)
        pages_and_texts.append({"page_number": page_number+1,
                                "page_char_count": len(text),
                                "page_word_count": len(text.split(" ")),
                                "page_sentence_count_raw": len(text.split(". ")),
                                "page_token_count": len(text) / 4,  # 1 token =4 chars
                                "text": formatted_text})
  return pages_and_texts
pages_and_texts=open_and_read_pdf(doc_path)

1208it [00:01, 618.30it/s]


In [ ]:
import pandas as pd

df = pd.DataFrame(pages_and_texts)
df.head()

,page_number,page_char_count,page_word_count,page_sentence_count_raw,page_token_count,text
0,1,31,5,1,7.75,human nutrition: 2020 edition
1,2,0,1,1,0.00,
2,3,322,43,1,80.50,human nutrition: 2020 edition university of ha...
3,4,214,31,2,53.50,human nutrition: 2020 edition by university of...
4,5,799,115,2,199.75,contents preface university of hawai‘i at māno...


In [ ]:
avg_token_count=df['page_token_count'].mean()
print('The avg nb of tokens in a page is: ',avg_token_count)

The avg nb of tokens in a page is:  287.25289735099335


In [ ]:
max_token_count=df['page_token_count'].max()
print('The max nb of tokens in a page is: ',max_token_count)

The max nb of tokens in a page is:  577.25


In [ ]:
from spacy.lang.en import English

def break_into_sentences_scipy(pages_and_texts):
  nlp = English()
  nlp.add_pipe("sentencizer") #component that helps detect sentence boundaries
  for item in tqdm(pages_and_texts):
      item["sentences"] = list(nlp(item["text"]).sents)

      # Make sure all sentences are strings
      item["sentences"] = [str(sentence) for sentence in item["sentences"]]

      # Count the sentences
      item["page_sentence_count_spacy"] = len(item["sentences"])
break_into_sentences_scipy(pages_and_texts)

100%|██████████| 1208/1208 [00:01<00:00, 933.11it/s] 


In [ ]:
df = pd.DataFrame(pages_and_texts)
df.head()

,page_number,page_char_count,page_word_count,page_sentence_count_raw,page_token_count,text,sentences,page_sentence_count_spacy
0,1,31,5,1,7.75,human nutrition: 2020 edition,[human nutrition: 2020 edition],1
1,2,0,1,1,0.00,,[],0
2,3,322,43,1,80.50,human nutrition: 2020 edition university of ha...,[human nutrition: 2020 edition university of h...,1
3,4,214,31,2,53.50,human nutrition: 2020 edition by university of...,[human nutrition: 2020 edition by university o...,1
4,5,799,115,2,199.75,contents preface university of hawai‘i at māno...,[contents preface university of hawai‘i at mān...,2


In [ ]:
avg_page_sentence_count_spacy=df['page_sentence_count_spacy'].mean()
print('The avg_page_sentence_count_spacy is: ',avg_page_sentence_count_spacy)

The avg_page_sentence_count_spacy is:  10.350165562913908


On average each of our pages has 10 sentences.
And an average total of 287 tokens per page.
So our groups of 10 sentences will also be ~287 tokens long.

In [ ]:
sentence_chunk_size = 10

def split_into_chunks(pages_and_texts):
  def split_list(input_list: list, sentence_chunk_size: int):
    return [input_list[i:i + sentence_chunk_size] for i in range(0, len(input_list), sentence_chunk_size)]

  for item in tqdm(pages_and_texts):
      item["sentence_chunks"] = split_list(input_list=item["sentences"],sentence_chunk_size=sentence_chunk_size)
      item["num_chunks"] = len(item["sentence_chunks"])
split_into_chunks(pages_and_texts)

100%|██████████| 1208/1208 [00:00<00:00, 342971.59it/s]


In [ ]:
df = pd.DataFrame(pages_and_texts)
df.head()

,page_number,page_char_count,page_word_count,page_sentence_count_raw,page_token_count,text,sentences,page_sentence_count_spacy,sentence_chunks,num_chunks
0,1,31,5,1,7.75,human nutrition: 2020 edition,[human nutrition: 2020 edition],1,[[human nutrition: 2020 edition]],1
1,2,0,1,1,0.00,,[],0,[],0
2,3,322,43,1,80.50,human nutrition: 2020 edition university of ha...,[human nutrition: 2020 edition university of h...,1,[[human nutrition: 2020 edition university of ...,1
3,4,214,31,2,53.50,human nutrition: 2020 edition by university of...,[human nutrition: 2020 edition by university o...,1,[[human nutrition: 2020 edition by university ...,1
4,5,799,115,2,199.75,contents preface university of hawai‘i at māno...,[contents preface university of hawai‘i at mān...,2,[[contents preface university of hawai‘i at mā...,1


In [ ]:
import re

def create_chunks(pages_and_texts):
  chunks = []
  for item in tqdm(pages_and_texts):
      for sentence_chunk in item["sentence_chunks"]:
          chunk_dict = {}
          chunk_dict["page_number"] = item["page_number"]

          # Join the sentences together into a paragraph-like structure, aka a chunk (so they are a single string)
          joined_sentence_chunk = "".join(sentence_chunk).replace("  ", " ").strip()
          joined_sentence_chunk = re.sub(r'\.([A-Z])', r'. \1', joined_sentence_chunk) # ".A" -> ". A" for any full-stop/capital letter combo
          chunk_dict["sentence_chunk"] = joined_sentence_chunk

          # Get stats about the chunk
          chunk_dict["chunk_char_count"] = len(joined_sentence_chunk)
          chunk_dict["chunk_word_count"] = len([word for word in joined_sentence_chunk.split(" ")])
          chunk_dict["chunk_token_count"] = len(joined_sentence_chunk) / 4 # 1 token = ~4 characters

          chunks.append(chunk_dict)
  return chunks
chunks=create_chunks(pages_and_texts)

100%|██████████| 1208/1208 [00:00<00:00, 39749.26it/s]


In [ ]:
len(chunks)

1845

In [ ]:
chunks_df = pd.DataFrame(chunks)
chunks_df.head()

,page_number,sentence_chunk,chunk_char_count,chunk_word_count,chunk_token_count
0,1,human nutrition: 2020 edition,29,4,7.25
1,3,human nutrition: 2020 edition university of ha...,308,42,77.00
2,4,human nutrition: 2020 edition by university of...,210,30,52.50
3,5,contents preface university of hawai‘i at māno...,765,115,191.25
4,6,lifestyles and nutrition university of hawai‘i...,939,142,234.75


Assuming that sentences with a maximum of 20 tokens(=15 words) re irrelevnt (page titles, etc...) let's filer them out.

In [ ]:
min_token_length = 20
chunks= chunks_df[chunks_df["chunk_token_count"] > min_token_length].to_dict(orient="records")

{'page_number': 3,
 'sentence_chunk': 'human nutrition: 2020 edition university of hawai‘i at mānoa food science and human nutrition program alan titchenal, skylar hara, noemi arceo caacbay, william meinke-lau, ya-yun yang, marie kainoa fialkowski revilla, jennifer draper, gemady langfelder, cheryl gibby, chyna nicole chun, and allison calabrese',
 'chunk_char_count': 308,
 'chunk_word_count': 42,
 'chunk_token_count': 77.0}

In [ ]:
len(chunks)

1748

# Embedding the Data

In [ ]:
from sentence_transformers import SentenceTransformer
embedding_model = SentenceTransformer(model_name_or_path="all-mpnet-base-v2",
                                      device="cpu") # choose the device to load the model to (note: GPU will often be *much* faster than CPU)

# Create a list of sentences to turn into numbers
sentences = [
    "The Sentences Transformers library provides an easy and open-source way to create embeddings.",
    "Sentences can be embedded one by one or as a list of strings.",
    "Embeddings are one of the most powerful concepts in machine learning!",
    "Learn to use embeddings well and you'll be well on your way to being an AI engineer."
]

# Sentences are encoded/embedded by calling model.encode()
embeddings = embedding_model.encode(sentences)
print(embeddings.shape)

/usr/local/lib/python3.10/dist-packages/sentence_transformers/cross_encoder/CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]